File: PerformacApp/Sources/Performac/Core/Collectors.swift
// Core/Collectors.swift — pure string→object parsers for every CLI output.
// Port of v1 collectors.js: bad input → nil/[]/empty state; never throws.
import Foundation

struct PsRow: Equatable, Sendable {
    var pid: Int64
    var cpu: Double
    var rssMb: Int64
    var name: String
}

func parsePs(_ text: String?) -> [PsRow] {
    var rows: [PsRow] = []
    for line in (text ?? "").split(separator: "\n") {
        let s = String(line)
        guard let m = s.wholeMatch(of: /^\s*(\d+)\s+([\d.]+)\s+(\d+)\s+(.+?)\s*$/) else { continue }
        rows.append(PsRow(
            pid: Int64(m.1) ?? 0,
            cpu: Double(m.2) ?? 0,
            rssMb: Int64(((Double(m.3) ?? 0) / 1024).rounded()),
            name: m.4.trimmingCharacters(in: .whitespaces)))
    }
    return rows
}

struct DfRow: Equatable, Sendable {
    var mount: String
    var freeGb: Double
    var totalGb: Double
}

func parseDf(_ text: String?) -> [DfRow] {
    var rows: [DfRow] = []
    for line in (text ?? "").split(separator: "\n") {
        let f = line.split(whereSeparator: { $0.isWhitespace }).map(String.init)
        if f.count < 9 || !f[0].hasPrefix("/dev/disk") { continue }
        rows.append(DfRow(
            mount: f[8...].joined(separator: " "),
            freeGb: (Double(f[3]) ?? 0) / 1_048_576,
            totalGb: (Double(f[1]) ?? 0) / 1_048_576))
    }
    return rows
}

func parseFrontAppName(_ text: String?) -> String? {
    guard let m = (text ?? "").wholeMatch(of: /"LSDisplayName"\s*=\s*"([^"]*)"/) else { return nil }
    return String(m.1)
}

struct ThermState: Equatable, Sendable {
    var available: Bool
    var cpuSpeedLimit: Int64?
}

func parseTherm(_ text: String?) -> ThermState {
    if let m = (text ?? "").wholeMatch(of: /CPU_Speed_Limit\s*=\s*(\d+)/) {
        return ThermState(available: true, cpuSpeedLimit: Int64(m.1))
    }
    return ThermState(available: false, cpuSpeedLimit: nil)
}

func parseThermlogLine(_ line: String?) -> Int64? {
    guard let m = (line ?? "").firstMatch(of: /Thermal Warning Level\s*=\s*(\d+)/) else { return nil }
    return Int64(m.1)
}

struct BatteryState: Equatable, Sendable {
    var cycleCount: Int64
    var designCap: Int64
    var nominalCap: Int64
    var healthPct: Double
}

func parseBattery(_ text: String?) -> BatteryState? {
    func g(_ k: String) -> Int64? {
        let re = try! NSRegularExpression(pattern: "\"" + k + "\"\\s*=\\s*(\\d+)")
        let str = text ?? ""
        guard let m = re.firstMatch(in: str, range: NSRange(str.startIndex..., in: str)),
              m.range(at: 1).location != NSNotFound else { return nil }
        return Int64((str as NSString).substring(with: m.range(at: 1)))
    }
    guard let nominal = g("NominalChargeCapacity"),
          let design = g("DesignCapacity"),
          let cycles = g("CycleCount") else { return nil }
    let pct = (Double(nominal) / Double(design) * 1000).rounded() / 10
    return BatteryState(cycleCount: cycles, designCap: design, nominalCap: nominal, healthPct: pct)
}

struct TmDestinations: Equatable, Sendable {
    var configured: Bool
    var names: [String]
}

func parseTmDestinations(_ text: String?) -> TmDestinations {
    var names: [String] = []
    for line in (text ?? "").split(separator: "\n") {
        if let m = line.wholeMatch(of: /^Name\s*:\s*(.+)$/) { names.append(String(m.1).trimmingCharacters(in: .whitespaces)) }
    }
    return TmDestinations(configured: !names.isEmpty, names: names)
}

struct TmLatest: Equatable, Sendable {
    var backupISO: String
}

func parseTmLatest(_ text: String?) -> TmLatest? {
    guard let m = (text ?? "").firstMatch(of: /(\d{4})-(\d{2})-(\d{2})-(\d{6})\.backup/) else { return nil }
    let t = m.4
    let iso = "\(m.1)-\(m.2)-\(m.3)T\(t.prefix(2)):\(t.dropFirst(2).prefix(2)):\(t.dropFirst(4).prefix(2))"
    return TmLatest(backupISO: iso)
}

struct ResolveConfig: Equatable, Sendable {
    var fsRoot: String?
    var cacheDir: String?
}

func parseResolveConfig(_ text: String?) -> ResolveConfig {
    let s = text ?? ""
    let root = s.firstMatch(of: /(?m)^Site\.\d+\.FS\.1\.Root\s*=\s*(.+)$/)
    let dir = s.firstMatch(of: /(?m)^RenderCaching\.CacheDir\s*=\s*(.+)$/)
    return ResolveConfig(
        fsRoot: root.map { String($0.1).trimmingCharacters(in: .whitespaces) },
        cacheDir: dir.map { String($0.1).trimmingCharacters(in: .whitespaces) })
}

// real `diskutil activity` lines carry the literal string DAVolumeName = '<null>' for unnamed
// disks (disk/container/scheme events) — that is NOT a name, return nil so no event is written.
// ts comes from the line's own Time=YYYYMMDD-HH:MM:SS (local), nil when absent.
struct DiskActivity: Equatable, Sendable {
    var kind: String   // appeared | disappeared
    var volume: String
    var ts: Int64?
}

func parseDiskutilActivity(_ line: String?) -> DiskActivity? {
    let s = line ?? ""
    guard let m = s.firstMatch(of: /\*\*\*Disk(Appeared|Disappeared).*DAVolumeName = '([^']*)'/) else { return nil }
    let name = String(m.2)
    if name.isEmpty || name == "<null>" { return nil }
    var ts: Int64?
    if let t = s.firstMatch(of: /Time=(\d{4})(\d{2})(\d{2})-(\d{2}):(\d{2}):(\d{2})/) {
        var comps = DateComponents()
        comps.year = Int(t.1); comps.month = Int(t.2); comps.day = Int(t.3)
        comps.hour = Int(t.4); comps.minute = Int(t.5); comps.second = Int(t.6)
        if let date = Calendar.current.date(from: comps) { ts = Int64(date.timeIntervalSince1970 * 1000) }
    }
    return DiskActivity(kind: String(m.1).lowercased(), volume: name, ts: ts)
}

// `diskutil info -plist <vol>` — booleans render as <true/>/<false/> tags
struct DiskInfo: Equatable, Sendable {
    var `internal`: Bool
    var ejectable: Bool
}

func parseDiskutilInfo(_ text: String?) -> DiskInfo? {
    let s = text ?? ""
    func val(_ k: String) -> Bool? {
        let re = try! NSRegularExpression(pattern: "<key>" + k + "</key>\\s*<(true|false)/>")
        guard let m = re.firstMatch(in: s, range: NSRange(s.startIndex..., in: s)),
              m.range(at: 1).location != NSNotFound else { return nil }
        return (s as NSString).substring(with: m.range(at: 1)) == "true"
    }
    let internalV = val("Internal")
    let ejectable = val("Ejectable")
    if internalV == nil && ejectable == nil { return nil }
    return DiskInfo(internal: internalV ?? false, ejectable: ejectable ?? false)
}

// `pmset -g batt -o` → "Now drawing from 'AC Power'" / "'Battery Power'"
func parsePower(_ text: String?) -> String? {
    guard let m = (text ?? "").wholeMatch(of: /Now drawing from '([^']+)'/) else { return nil }
    return String(m.1)
}

func parseLoginItems(_ text: String?) -> [String] {
    (text ?? "").split(separator: ",").map { String($0).trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
}

File: collectors.js
// collectors.js — pure string→object parsers for every CLI output.
// No child_process in this file. Bad input → null/[]/empty state; never throws.
export function parsePs(text) {
  const rows = [];
  for (const line of String(text ?? '').split('\n')) {
    const m = line.match(/^\s*(\d+)\s+([\d.]+)\s+(\d+)\s+(.+?)\s*$/);
    if (!m) continue;
    rows.push({ pid: Number(m[1]), cpu: Number(m[2]), rssMb: Math.round(Number(m[3]) / 1024), name: m[4].trim() });
  }
  return rows;
}

export function parseDf(text) {
  const rows = [];
  for (const line of String(text ?? '').split('\n')) {
    const f = line.trim().split(/\s+/);
    if (f.length < 9 || !f[0].startsWith('/dev/disk')) continue;
    rows.push({ mount: f.slice(8).join(' '), freeGb: Number(f[3]) / 1048576, totalGb: Number(f[1]) / 1048576 });
  }
  return rows;
}

export function parseFrontAppName(text) {
  const m = String(text ?? '').match(/"LSDisplayName"\s*=\s*"([^"]*)"/);
  return m ? m[1] : null;
}

export function parseTherm(text) {
  const m = String(text ?? '').match(/CPU_Speed_Limit\s*=\s*(\d+)/);
  return m ? { available: true, cpuSpeedLimit: Number(m[1]) } : { available: false, cpuSpeedLimit: null };
}

export function parseThermlogLine(line) {
  const m = String(line ?? '').match(/Thermal Warning Level\s*=\s*(\d+)/);
  return m ? { level: Number(m[1]) } : null;
}

export function parseBattery(text) {
  const g = k => {
    const m = String(text ?? '').match(new RegExp('"' + k + '"\\s*=\\s*(\\d+)'));
    return m ? Number(m[1]) : null;
  };
  const nominal = g('NominalChargeCapacity');
  const design = g('DesignCapacity');
  const cycles = g('CycleCount');
  if (nominal == null || design == null || cycles == null) return null;
  return { cycleCount: cycles, designCap: design, nominalCap: nominal, healthPct: Math.round((nominal / design) * 1000) / 10 };
}

export function parseTmDestinations(text) {
  const names = [...String(text ?? '').matchAll(/^Name\s*:\s*(.+)$/gm)].map(m => m[1].trim());
  return { configured: names.length > 0, names };
}

export function parseTmLatest(text) {
  const m = String(text ?? '').match(/(\d{4})-(\d{2})-(\d{2})-(\d{6})\.backup/);
  if (!m) return null;
  const t = m[4];
  return { backupISO: `${m[1]}-${m[2]}-${m[3]}T${t.slice(0, 2)}:${t.slice(2, 4)}:${t.slice(4, 6)}` };
}

export function parseResolveConfig(text) {
  const s = String(text ?? '');
  const root = s.match(/^Site\.\d+\.FS\.1\.Root\s*=\s*(.+)$/m);
  const dir = s.match(/^RenderCaching\.CacheDir\s*=\s*(.+)$/m);
  return { fsRoot: root ? root[1].trim() : null, cacheDir: dir ? dir[1].trim() : null };
}

// real `diskutil activity` lines carry the literal string DAVolumeName = '<null>' for unnamed
// disks (disk/container/scheme events) — that is NOT a name, return null so no event is written.
// ts comes from the line's own Time=YYYYMMDD-HH:MM:SS (local), null when absent.
export function parseDiskutilActivity(line) {
  const m = String(line ?? '').match(/\*\*\*Disk(Appeared|Disappeared).*DAVolumeName = '([^']*)'/);
  if (!m || !m[2] || m[2] === '<null>') return null;
  const t = String(line).match(/Time=(\d{4})(\d{2})(\d{2})-(\d{2}):(\d{2}):(\d{2})/);
  const ts = t ? new Date(Number(t[1]), Number(t[2]) - 1, Number(t[3]), Number(t[4]), Number(t[5]), Number(t[6])).getTime() : null;
  return { kind: m[1].toLowerCase(), volume: m[2], ts };
}

// `diskutil info -plist <vol>` — booleans render as <true/>/<false/> tags
export function parseDiskutilInfo(text) {
  const s = String(text ?? '');
  const val = k => {
    const m = s.match(new RegExp('<key>' + k + '</key>\\s*<(true|false)/>'));
    return m ? m[1] === 'true' : null;
  };
  const internal = val('Internal');
  const ejectable = val('Ejectable');
  if (internal == null && ejectable == null) return null;
  return { internal: internal ?? false, ejectable: ejectable ?? false };
}

// `pmset -g batt -o` → "Now drawing from 'AC Power'" / "'Battery Power'"
export function parsePower(text) {
  const m = String(text ?? '').match(/Now drawing from '([^']+)'/);
  return m ? m[1] : null;
}

export function parseLoginItems(text) {
  return String(text ?? '').split(',').map(s => s.trim()).filter(Boolean);
}
